### ISSUE TITLE
Fluent API configuration

### ISSUE THREAD

**Issue by wojtpl2** [wojtpl2/ExtendedXmlSerializer#109]
Your new fluent API isn't really fluent api. Eg:
```C#
new ConfigurationContainer()
  .ConfigureType<TestClassWithEncryptedData>()
    .Member(p => p.Salary).Encrypt()
  .Owner.Configuration
  .ConfigureType<TestClassPrimitiveTypesNullable>()
    .Member(p=>p.PropChar).Encrypt()
  .Owner.Configuration
  .UseEncryptionAlgorithm()
  .Create();
```
In my last API it was easier to do this:
```C#
new ConfigurationContainer()
  .ConfigureType<TestClassWithEncryptedData>()
    .Member(p => p.Salary).Encrypt()
  .ConfigureType<TestClassPrimitiveTypesNullable>()
    .Member(p=>p.PropChar).Encrypt()
  .UseEncryptionAlgorithm()
  .Create();
```

In your solution all functions are public eg:
`configuration.Find<EncryptionExtension>();`
or 
`configuration.Add<Object>()` <-- you are really able to call this 

In my only functions to configuration were public. All functions to internal use were internal. In your solution all function are public.

You can get my last API and you can check:
`git checkout bca8b59a8f04e181a72b636d26597e1fc942738f`

For me if you type `.` in fluent api you should be able to call all functions from your level and all higher level.

What do you think? 
I wanted to do this but it is very difficult to change in your solution.

**Comment by Mike-EEE** [comment]
Yeah it was my first stab at a fluent API.  Like I mentioned earlier (WAY earlier LOL) I was not very happy with it... like I would rate it a 6/10.  I am happy w/ everything else, though. 😄 

Maybe we make a v2.5 issue to improve fluent API?

**Comment by wojtpl2** [comment]
Why don't we use my old Api instead of your? I know that my old Api need many improvements and other changes. I know that my solution wasn't be perfect, but I think it was better then currently API. 
What do you think? 
I'm afraid that we'll never do v2.5 ;)

**Comment by Mike-EEE** [comment]
LOL... I know, I sound like a blasted project manager kicking the can down the road at the moment.  I totally understand your apprehension.

When you say "old" API do you mean the one that was designed?

FWIW I tried to stick to it as much as possible, but there were some issues that I encountered and we ended up w/ what I have.  Again, I am not happy with it either!

BTW, the above can be reduced to:

```
new ConfigurationContainer()
  .ConfigureType<TestClassWithEncryptedData>()
    .Member(p => p.Salary).Encrypt()
  .Configuration
  .ConfigureType<TestClassPrimitiveTypesNullable>()
    .Member(p=>p.PropChar).Encrypt()
  .Configuration
  .UseEncryptionAlgorithm()
  .Create();
```

(this might not help much but thought I would mention)

After this week I will be 100% on this sucker again until July.  Believe me when I say this will get done!

**Comment by wojtpl2** [comment]
Old api in V2 is there:
`git checkout bca8b59a8f04e181a72b636d26597e1fc942738f` 
For users it was perfect - They could only call what they need. All function that was used by us was internal. Now all function are public. 

We should release version with good api - it shouldn't change any more.

**Comment by Mike-EEE** [comment]
Ah ok... I will have to investigate this.  Keep in mind that a lot of code is public until #70 is done.

I am also open to making this part of v2.0 if you can wait a few more days. 👼

**Comment by wojtpl2** [comment]
We can wait. I would like to release good version.

This is really difficult task.
```C#
	public interface IConfigurationContainer : ICollection<ISerializerExtension>
	{
		T Find<T>() where T : ISerializerExtension;

		IExtendedXmlSerializer Create();
	}
```
This interface should be split into two different:
```C#
	public interface IConfigurationContainer
	{
		IExtendedXmlSerializer Create();
	}
	internal interface IConfigurationContainerInternal : ICollection<ISerializerExtension>
	{
		T Find<T>() where T : ISerializerExtension;
	}
```
Public interface can have only functions for user.
Internal interface has to have all functions which are for us.  Inherit from ICollection is for us, not for user. User can't call function from interface ICollection. 
The most difficult place will be extension methods. I think that this methods should be moved to interface.

I can do this task but I can take me week or two. Do you want do this?

**Comment by Mike-EEE** [comment]
Ahhh yes that is fair feedback.  I really like your suggestion and can't think I thought of that. :P  Also, I hate generic interface methods, so should have thought of making that internal as well.  

I will have to revisit the extension methods again as I remember there were some key decisions I made around them.  But again, my first time doing Fluent. :)

Do you have a good resource around Fluent that I can check out?  That might help, but no promises.

I agree this is better done sooner than later, since it is public-facing.  I can make this part of 2.0, but it will push it back a few days (but sounds like it is OK with you).  I will adjust schedule.

**Comment by wojtpl2** [comment]
Your api isn't wrong but for internal use. In public library the the most important is interface API. 

I don't have any good resource. I learned it from open source projects. FluentValidation was my first touch with fluent API. https://github.com/JeremySkinner/FluentValidation
I read Moq, Autofac, automapper too.

**Comment by Mike-EEE** [comment]
OK... I've committed in the new ConfigurationAPI, and have gotten rid of a lot of the extension methods.  It's not perfect, but it is more straight-forward and allows you to do the above.  Or at least should.  All tests are passing.  So please take a look at this and if it looks OK to you, go ahead and release alpha 7.

If there are any significant changes that are needed, I would recommend moving them out to 2.1.  I am not sure how much I will be able to do in this regard, however.  It might take another pass or two to get this as desired, but I think it should do for now and v2. :)

**Comment by wojtpl2** [comment]
I look over new API. It looks good. I need to more time to read it and understand all. 

I've pushed one commit with updating documentation.

**Comment by Mike-EEE** [comment]
Cooooool... I forgot to update the milestones with documentation goals.  That's what I will be working on for the next few days.  I am going on vacation in a week so it has to be done before then. 😄

**Comment by Mike-EEE** [comment]
This issue has been addressed.  Please feel free to open a new issue if there is anything remaining/expected that persists.
